### PR TITLE
Mark [...a=0] as an invalid assignment target

### LINKED ISSUE #355: "Error: Not reached" when trying to parse `(([...a=1]))`
```
> require("shift-parser").parseScript("(([...a=1]))")
Error: Not reached
```

### PR DESCRIPTION
**Pull request by bakkot**
Fixes #355.

### CODE REVIEW

File: src/parser.js [hunk 1]
```diff
@@ -1892,6 +1892,9 @@ export class GenericParser extends Tokenizer {
             rest = expr;
             break;
           }
+          if (expr.type !== 'ArrayExpression' && expr.type !== 'ObjectExpression' && !isValidSimpleAssignmentTarget(expr)) {
```
Comment (bakkot) [inline src/parser.js:1895]: The only thing I can think of such that this criteria is satisfied but we would not already have recognized that we can't be parsing an assignment target is `AssignmentExpression`, so it's possible we could replace this condition with `if (expr.type === 'AssignmentExpression')`.

On the other hand, this is closer to [what the grammar says](https://tc39.github.io/ecma262/#sec-destructuring-assignment-static-semantics-early-errors).

File: test/expressions/array-expression.js [hunk 2]
```diff
@@ -62,7 +62,19 @@ suite('Parser', function () {
 
   });
 
-  testParse('[a, ...(b=c)]', expr, {
```
Comment (michaelficarra) [inline test/expressions/array-expression.js:65]: If this should fail to parse, please assert that.

Comment (bakkot) [reply test/expressions/array-expression.js:65]: It shouldn't. It's valid code and continues to parse; it's just that the parentheses aren't really doing anything interesting except making the parser's job easier.

Comment (michaelficarra) [reply test/expressions/array-expression.js:65]: Haha oh man, this is all an expression. I got it.

**Review by michaelficarra — APPROVED** [review]

**Pull request merged**